Core/src/Geometry/Layer.cpp

```diff
@@ -119,16 +120,18 @@ Acts::Layer::compatibleSurfaces(
     return sIntersections;
   }
 
+  double nearLimit = options.nearLimit;
+  double farLimit = options.farLimit;
+
+  // TODO this looks like a major hack; is this really needed?
   // (0) End surface check
   // @todo: - we might be able to skip this by use of options.pathLimit
   // check if you have to stop at the endSurface
-  double nearLimit = options.nearLimit;
-  double farLimit = options.farLimit;
   if (options.endObject != nullptr) {
     // intersect the end surface
     // - it is the final one don't use the boundary check at all
     SurfaceIntersection endInter =
-        options.endObject
+        static_cast<const Surface*>(options.endObject)
```

Comment:
Here we do in fact cast back the void pointer. Can we add a struct with a type index and a cast function?